src/crypto/crypto_dh.cc

```diff
@@ -539,62 +540,73 @@ WebCryptoKeyExportStatus DHKeyExportTraits::DoExport(
-      EVP_PKEY_derive(ctx.get(), nullptr, &out_size) <= 0)
-    return AllocatedBuffer();
+      EVP_PKEY_derive(ctx.get(), nullptr, out_size) <= 0)
+    return false;
 
-  AllocatedBuffer result = AllocatedBuffer::AllocateManaged(env, out_size);
-  CHECK_NOT_NULL(result.data());
+  *buf = MallocOpenSSL<char>(*out_size);
+  size_t remainder_size = *out_size;
+  if (EVP_PKEY_derive(ctx.get(),
+                      reinterpret_cast<unsigned char*>(*buf),
+                      &remainder_size) <= 0) {
+    OPENSSL_clear_free(buf, *out_size);
+    return false;
+  }
 
-  unsigned char* data = reinterpret_cast<unsigned char*>(result.data());
-  if (EVP_PKEY_derive(ctx.get(), data, &out_size) <= 0)
-    return AllocatedBuffer();
+  ZeroPadDiffieHellmanSecret(remainder_size, *buf, *out_size);
 
-  ZeroPadDiffieHellmanSecret(out_size, &result);
-  return result;
+  return true;
 }
 
-// The version of StatelessDiffieHellman that returns an AllocatedBuffer
-// is not threadsafe because of the AllocatedBuffer allocation of a
-// v8::BackingStore (it'll cause much crashing if we call it from a
-// libuv worker thread). This version allocates a ByteSource instead,
-// which we can convert into a v8::BackingStore later.
-// TODO(@jasnell): Eliminate the code duplication between these two
-// versions of the function.
-ByteSource StatelessDiffieHellmanThreadsafe(
+Local<Value> StatelessDiffieHellman(
     Environment* env,
-    ManagedEVPPKey our_key,
-    ManagedEVPPKey their_key) {
+    const ManagedEVPPKey& our_key,
+    const ManagedEVPPKey& their_key) {
+  char* buf = nullptr;
   size_t out_size;
-
-  EVPKeyCtxPointer ctx(EVP_PKEY_CTX_new(our_key.get(), nullptr));
-  if (!ctx ||
-      EVP_PKEY_derive_init(ctx.get()) <= 0 ||
-      EVP_PKEY_derive_set_peer(ctx.get(), their_key.get()) <= 0 ||
-      EVP_PKEY_derive(ctx.get(), nullptr, &out_size) <= 0)
-    return ByteSource();
-
-  char* buf = MallocOpenSSL<char>(out_size);
-  ByteSource out = ByteSource::Allocated(buf, out_size);
-
-  if (EVP_PKEY_derive(
-          ctx.get(),
-          reinterpret_cast<unsigned char*>(buf),
-          &out_size) <= 0) {
+  if (!CreateStatelessDiffieHellmanContext(our_key,
+                                           their_key,
+                                           &buf,
+                                           &out_size))
+    return Undefined(env->isolate());
+
+  void* hint = reinterpret_cast<void*>(static_cast<uintptr_t>(out_size));
+  return Buffer::New(env,
+                     buf,
+                     out_size,
+                     [](char* data, void* hint) {
+                       size_t out_size = static_cast<size_t>(
+                           reinterpret_cast<uintptr_t>(hint));
+                       OPENSSL_clear_free(data, out_size);
+                     },
+                     hint).FromMaybe(Local<Object>());
+}
+
+// The version of StatelessDiffieHellman that returns a Buffer
```

Comment:
@addaleax do you know why calling a function that allocates a `v8::BackingStore` from a `uv_thread_t` causes a crash?

Reply:
@RaisinTen Yeah, creating `v8::BackingStore` is unfortunately not threadsafe, it requires having acquired the lock for the Isolate in question (unfortunately)

Reply:
@addaleax In that case, instead of having two functions do essentially the same thing, wouldn't it be better to just have a single function return a `ByteSource` because it is threadsafe? In the function where we need a `Buffer`, we can just convert it from a `Local<ArrayBuffer>` by passing it through `ByteSource::ToArrayBuffer` first, right?

Reply:
@RaisinTen That seems fine to me, yes :+1:

Reply:
@addaleax done, PTAL